worldedit-core/src/main/java/com/sk89q/worldedit/command/GenerationCommands.java

```diff
@@ -122,67 +122,92 @@ public void cyl(Player player, LocalSession session, EditSession editSession, Pa
-            "you can generate an ellipsoid. The order of the ellipsoid radii\n" +
-            "is north/south, up/down, east/west.",
-        min = 2,
-        max = 3
+            aliases = { "/sphere" },
+            usage = "<block> <radius>[,<radius>,<radius>] [raised?]",
+            flags = "hsip",
+            desc = "Generates a filled sphere.",
+            help =
+                    "Generates a filled sphere.\n" +
+                            "By specifying 3 radii, separated by commas,\n" +
+                            "you can generate an ellipsoid. The order of the ellipsoid radii\n" +
+                            "is north/south, up/down, east/west.",
+            min = 2,
+            max = 3
     )
     @CommandPermissions("worldedit.generation.sphere")
     @Logging(PLACEMENT)
-    public void sphere(Player player, LocalSession session, EditSession editSession, Pattern pattern, String radiusString, @Optional("false") boolean raised, @Switch('h') boolean hollow) throws WorldEditException {
+    public void sphere(Player player, LocalSession session, EditSession editSession, Pattern pattern, String radiusString, @Optional("false") boolean raised, @Switch('h') boolean hollow, @Switch('s') boolean hemi, @Switch('i') boolean upsideDown, @Switch('p') boolean precisionMode) throws WorldEditException {
         String[] radii = radiusString.split(",");
         final double radiusX, radiusY, radiusZ;
         switch (radii.length) {
-        case 1:
-            radiusX = radiusY = radiusZ = Math.max(1, Double.parseDouble(radii[0]));
-            break;
-
-        case 3:
-            radiusX = Math.max(1, Double.parseDouble(radii[0]));
-            radiusY = Math.max(1, Double.parseDouble(radii[1]));
-            radiusZ = Math.max(1, Double.parseDouble(radii[2]));
-            break;
-
-        default:
-            player.printError("You must either specify 1 or 3 radius values.");
-            return;
+            case 1:
+                radiusX = radiusY = radiusZ = Math.max(1, Double.parseDouble(radii[0]));
+                break;
+
+            case 3:
+                radiusX = Math.max(1, Double.parseDouble(radii[0]));
+                radiusY = Math.max(1, Double.parseDouble(radii[1]));
+                radiusZ = Math.max(1, Double.parseDouble(radii[2]));
+                break;
+
+            default:
+                player.printError("You must either specify 1 or 3 radius values.");
+                return;
         }
 
         worldEdit.checkMaxRadius(radiusX);
         worldEdit.checkMaxRadius(radiusY);
         worldEdit.checkMaxRadius(radiusZ);
 
         Vector pos = session.getPlacementPosition(player);
-        if (raised) {
+        if (raised && !hemi) {
             pos = pos.add(0, radiusY, 0);
         }
 
-        int affected = editSession.makeSphere(pos, Patterns.wrap(pattern), radiusX, radiusY, radiusZ, !hollow);
+        Vector directionVector = null;
+        if (hemi && precisionMode) {
```

Comment:
One note is that if I do //sphere stone 4 -ip, it will only print the message related to the -p flag since it's tested for first. I can make it so it prints both, but it will clutter the method a little bit.